### ISSUE TITLE
PC-031 — XGBoost G(x) model stored as pickle of an older XGBoost version

### ISSUE THREAD

**Issue by itigges22** [itigges22/ATLAS#73]
**Status (at mirror time):** Fixed (rebuild needed)  •  **Severity:** Low (warning)

**One-liner:** Re-exported pickle as `gx_xgboost.json` via Booster.save_model; loader prefers JSON, falls back to pickle. No more startup warning

---

- **Status:** Open
- **Severity:** Low (warning at startup, model still loads correctly)
- **Location:** `geometric-lens/geometric_lens/service.py:80-89`,
  `geometric-lens/geometric_lens/models/gx_xgboost.pkl`,
  `scripts/download-models.sh:32`

**Targeted rebuild after fix:** none (model file swap + service.py
edit; bind-mount via `docker-compose.override.yml.example` picks
up the change with `docker compose restart geometric-lens`)

### Symptom

Every geometric-lens startup logs:

```
/app/geometric_lens/service.py:89: UserWarning: WARNING:
…/error_msg.h:83: If you are loading a serialized model (like
pickle in Python, RDS in R) or configuration generated by an
older version of XGBoost, please export the model by calling
`Booster.save_model` from that version first, then load it back
in current version. See:
    https://xgboost.readthedocs.io/en/stable/tutorials/saving_model.html
…
  _gx_xgboost = pickle.load(f)
```

The very next log line confirms the model still works:

```
G(x) XGBoost loaded (AUC=0.8840, PCA 4096→128)
```

So functionally fine — but every operator sees a scary stack-style
warning at startup, and pickle compat is not guaranteed across
future XGBoost majors.

### Root cause

`gx_xgboost.pkl` was serialized using Python's `pickle.dump` against
an older `xgboost.sklearn.XGBClassifier`. The proxy's loader at
`service.py:84-89` does:

```python
import pickle
…
with open(xgb_path, 'rb') as f:
    _gx_xgboost = pickle.load(f)
```

Inside the container we have XGBoost 3.2.0; the pickle was written
under an earlier version. XGBoost emits the forward-compat warning
on every load. There's no way to silence it cleanly via warnings
filter — it's emitted from C++ via `_LIB.XGBLog`.

The right fix is to stop relying on pickle and use XGBoost's
native serialization (`save_model` → JSON), which is:
- version-stable (documented format),
- safer (no arbitrary-code-execution surface like pickle),
- smaller (typically ~30% smaller than pickle for the same model).

### Fix — proposed plan

Three artifacts:

1. **Re-export the model.** Run a one-shot script inside the
   geometric-lens container:
   ```python
   import pickle, xgboost as xgb
   with open('/app/geometric_lens/models/gx_xgboost.pkl', 'rb') as f:
       m = pickle.load(f)
   m.save_model('/app/geometric_lens/models/gx_xgboost.json')
   ```
   Then verify `predict_proba` matches the pickle output to
   numerical tolerance on a synthetic input batch.

2. **Update `service.py`.** Prefer the JSON file; fall back to
   pickle only if JSON is missing (one-release grace period for
   users who haven't refreshed their model dir):
   ```python
   import xgboost as xgb
   xgb_json = os.path.join(models_dir, "gx_xgboost.json")
   if os.path.exists(xgb_json):
       m = xgb.XGBClassifier()
       m.load_model(xgb_json)
       _gx_xgboost = m
   elif os.path.exists(xgb_path):  # legacy pickle
       with open(xgb_path, 'rb') as f:
           _gx_xgboost = pickle.load(f)
       logger.warning(
           "Loaded G(x) from legacy pickle. Re-export to "
           "gx_xgboost.json — this fallback will be removed."
       )
   ```

3. **Update `scripts/download-models.sh`.** Add `gx_xgboost.json`
   to `LENS_FILES` (line 32). Keep `gx_xgboost.pkl` in the list
   for the grace period; remove it once the JSON is published to
   the HF dataset and the fallback is dropped.

The HF-side step (uploading `gx_xgboost.json` to the
`itigges22/ATLAS` dataset alongside the pickle) is a manual
operation and not part of this code-side fix.

### Why not silence the warning?

XGBoost emits this from C++ via its logging layer; Python's
`warnings.filterwarnings("ignore")` doesn't catch it cleanly, and
even if it did, that only hides a real forward-compat risk. The
warning is doing its job; the right move is to fix the underlying
serialization, not muzzle the messenger.

### Service scope — targeted rebuild

If `docker-compose.override.yml` is active (PC-030 dev mode),
just:
```
docker compose restart geometric-lens
```

Without dev mode:
```
docker compose up -d --build geometric-lens
```

Either way: only `geometric-lens` is affected. No proxy, v3,
sandbox, or llama-server rebuild.

### How to verify

After the swap, geometric-lens startup should log:

```
G(x) XGBoost loaded (AUC=0.8840, PCA 4096→128)
```

…with **no** UserWarning preceding it. Confirm
`/score?query=...` still returns the same gx_score values it did
under the pickle path on a handful of fixed-seed inputs.

### Docs to update in the same change

- `geometric-lens/geometric_lens/models/README.md` — model file
  listing (replace `gx_xgboost.pkl` with `gx_xgboost.json`,
  document the legacy pickle as deprecated).
- `docs/ARCHITECTURE.md` — Lens/G(x) section if it references the
  pickle path explicitly.
- (No README, CONFIGURATION, TROUBLESHOOTING, API, or CLI
  changes.)

Files to touch (when fix lands):
- `geometric-lens/geometric_lens/service.py:78-108`
- `geometric-lens/geometric_lens/models/gx_xgboost.json` (new)
- `scripts/download-models.sh:28-35`
- `geometric-lens/geometric_lens/models/README.md`

### Fix (2026-04-30)

Re-exported the XGBoost model in native JSON format and switched
the loader to prefer it.

**Re-export.** Inside the running `geometric-lens` container,
loaded the legacy pickle, called `m.get_booster().save_model(...)`
to write `gx_xgboost.json`, copied out to host. Verified
bit-exact prediction equivalence: max diff `0.0` across 8
synthetic 128-dim inputs through `Booster.predict` against both
the pickled XGBClassifier's underlying booster and the
re-exported booster. New file is 1.49 MB vs 980 KB pickle —
slightly larger because the JSON dump includes more metadata,
but cheap and version-stable.

**Loader switch.** `geometric-lens/geometric_lens/service.py:80-110`:
- Prefer `gx_xgboost.json` via `xgb.Booster(); b.load_model(...)`.
- Fall back to `gx_xgboost.pkl` via `pickle.load` if JSON missing
  (one-release grace period). Logs `pickle (deprecated — re-export
  to gx_xgboost.json)` when this branch fires so users can spot
  the deprecation in logs.
- New `_BoosterClassifier` shim at `service.py:32-52` mimics
  `predict_proba(x)` over a raw `Booster`. For binary:logistic
  objectives, `Booster.predict` returns the positive-class
  probability directly, so the shim returns `[1-pos, pos]` to
  preserve the `proba[1]` indexing in the two existing call
  sites (`service.py:340, 410`). Avoids a hard dependency on
  scikit-learn at runtime.

**Side-effect:** the noisy `UserWarning: WARNING: …
error_msg.h:83` from XGBoost is gone for users on the JSON path.

Files touched:
- `geometric-lens/geometric_lens/service.py:32-52, 80-110`
- `geometric-lens/geometric_lens/models/gx_xgboost.json` (new)
- `geometric-lens/geometric_lens/models/README.md`
- `scripts/download-models.sh:28-36`

**Targeted rebuild:** `docker compose up -d --build geometric-lens`
(or just `docker compose restart geometric-lens` if dev-mode
override is active).

**HF dataset publishing of `gx_xgboost.json`** is a manual
follow-up (out of scope for this code change). Until that lands,
new installs that haven't pre-staged the JSON will use the
pickle fallback path.

---

---
*Mirrored from local `ISSUES.md`. ISSUES.md is gitignored; this issue is the canonical public record. PC-### numbering preserved for cross-references.*

**Comment by itigges22** [comment]
Closed: Fixed (rebuild needed). Shipped in dev branch commit d7a8980.
